### PR TITLE
Cancel a `ArmeriaServerCall` when an `HttpRequest` is closed.

### LINKED ISSUE #3452: `ClosedStreamException` triggered by `ArmeriaServerCall.doSendMessage()` not handled by Armeria
We got the following exception report from a user at our Slack workspace:

```
2021-04-12 14:14:02.858  WARN 8324 --- [-worker-nio-2-8] i.n.u.concurrent.AbstractEventExecutor   : A task raised an exception. Task: com.linecorp.armeria.common.RequestContext$$Lambda$1068/66855464@6d21f471
com.linecorp.armeria.common.stream.ClosedStreamException: null
	at com.linecorp.armeria.common.stream.ClosedStreamException.get(ClosedStreamException.java:37) ~[armeria-1.5.0.jar:na]
	at com.linecorp.armeria.common.stream.StreamWriter.write(StreamWriter.java:71) ~[armeria-1.5.0.jar:na]
	at com.linecorp.armeria.server.grpc.ArmeriaServerCall.doSendMessage(ArmeriaServerCall.java:287) ~[armeria-grpc-1.5.0.jar:na]
	at com.linecorp.armeria.server.grpc.ArmeriaServerCall.lambda$sendMessage$4(ArmeriaServerCall.java:270) ~[armeria-grpc-1.5.0.jar:na]
	at com.linecorp.armeria.common.RequestContext.lambda$makeContextAware$3(RequestContext.java:522) ~[armeria-1.5.0.jar:na]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164) ~[netty-common-4.1.55.Final.jar:4.1.55.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472) [netty-common-4.1.55.Final.jar:4.1.55.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497) [netty-transport-4.1.55.Final.jar:4.1.55.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.55.Final.jar:4.1.55.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.55.Final.jar:4.1.55.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.55.Final.jar:4.1.55.Final]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_281]
```

According to the user, the exception is triggered when a server attempts to write something to the stream cancelled by a client. The stack trace shows that we did not catch the exception but just left it propagated to Netty. We should instead let it go to a `RequestLog` or discard it if that makes sense.

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

An `HttpRequest` is canceled by remote peer and an `HttpResponseWriter`
is closed, the `ArmeriaServerCall` should also be closed with
`Status.CANCELLED`. However, there is a race condition between
`response.write()` and `call.close()`.
The closing signal progragates to `ArmeriaServerCall` using `res.whenComplete()`.
The cancel action is not directly invoked by the signal. It is submitted to
an event loop and rescheduled.
https://github.com/line/armeria/blob/480321970111ac6249775199fedadcc71b862e84/grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java#L195-L204

Modifications:

- Directly try to cancel a server call when a response is closed.
- Cancel a server call when a `HttpResponseWriter` is closed already.
- Correctly propagate a cancellation cause by calling `ServiceRequestContext.cancel(cause)` 

Result:

- You no longer see a `ClosedStreamException` when a gRPC call is cancelled
  by a client.
- Fixes #3452

### CODE REVIEW

**Comment by ikhoon** [comment]
@anuraaga Does it violate the gRPC error specification to close a server call with `Status.CANCELLED` if a client closes a stream after a half close?
https://github.com/grpc/grpc-java/blob/21269bd18864ee420678cf43891d93c553ebd9b8/stub/src/main/java/io/grpc/stub/ServerCallStreamObserver.java#L55-L56

>   * <p>This method may only be called during the initial call to the application, before the
>   * service returns its {@code StreamObserver}.
>
>   */
>  public abstract void setOnCancelHandler(Runnable onCancelHandler);

**Comment by ikhoon** [comment]
@anuraaga Gentle ping. Could you review this PR. I'd like to listen to your opinion on  `Status.CANCELLED` and half close 😆

File: core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java [hunk 1]
```diff
@@ -233,13 +233,10 @@ private void cleanup() {
...
             final ClosedSessionException cause = ClosedSessionException.get();
             unfinishedRequests.forEach((req, res) -> {
+                // An HTTP2 request is cancelled by Http2RequestDecoder.onRstStreamRead()
+                final boolean cancel = !protocol.isMultiplex();
                 // Mark the request stream as closed due to disconnection.
-                if (!protocol.isMultiplex()) {
-                    // An HTTP2 request is cancelled by Http2RequestDecoder.onRstStreamRead()
-                    req.ctx().cancel(cause);
-                }
-                req.close(cause);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:237]: Is it OK to remove this line?

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:237]: `req.abortResponse()` closes itself using `tryClose()`
https://github.com/line/armeria/blob/b9c203beff2958e8171eabb07f87b2631dace9e9/core/src/main/java/com/linecorp/armeria/server/DecodedHttpRequest.java#L162-L163

File: core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java [hunk 2]
```diff
@@ -446,7 +446,12 @@ public void run(Throwable cause) {
                 // This method will be invoked only when `canSchedule()` returns true.
                 assert state != State.DONE;
 
-                failAndRespond(cause, convertException(cause), Http2Error.INTERNAL_ERROR, true);
+                if (cause instanceof ClosedStreamException || cause instanceof ClosedSessionException) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java:449]: 👍

**Review by minwoox — APPROVED** [review]
Thanks!

**Comment by anuraaga** [comment]
@ikhoon I probably copied that test from upstream, but can't find it anymore since it was perhaps refactored in the several years. I suspect returning `CANCELLED` is ok too, gRPC responses should be returned where possible.

**Review by trustin — APPROVED** [review]
Thanks a lot, @ikhoon !

**Pull request merged**